File: webapp/src/types/ai.ts
// Mirrors backend/app/models/resume_analysis.py::AIJobStatus - shared by
// both ResumeAnalysis and AtsScore (see backend/app/models/ats_score.py's
// module docstring for why the two reuse one Postgres enum).
export type AIJobStatus = 'pending' | 'processing' | 'completed' | 'failed'

export const AI_JOB_STATUSES: readonly AIJobStatus[] = [
  'pending',
  'processing',
  'completed',
  'failed',
]

// --- Gemini structured-output shapes ---------------------------------------
// Mirrors backend/app/schemas/ai.py's ParsedResume/WorkExperienceItem/
// EducationItem/AtsScoreResult exactly (down to field presence - every
// field there is required, even where nullable, per that file's own
// docstring on Gemini's response_schema constraints; doesn't matter here
// since the frontend only ever reads these, never sends them).

export interface WorkExperienceItem {
  company: string
  title: string
  start_date: string | null
  end_date: string | null
  description: string | null
}

export interface EducationItem {
  institution: string
  degree: string | null
  field_of_study: string | null
  graduation_date: string | null
}

export interface ParsedResume {
  full_name: string | null
  email: string | null
  phone: string | null
  summary: string | null
  skills: string[]
  work_experience: WorkExperienceItem[]
  education: EducationItem[]
  total_years_experience: number | null
}

export interface AtsScoreResult {
  score: number
  summary: string
  matched_keywords: string[]
  missing_keywords: string[]
  suggestions: string[]
}

// --- Resume analyses ---------------------------------------------------------

export interface ResumeAnalysisCreatePayload {
  document_id: string
}

// Mirrors ResumeAnalysisRead. `parsed_data` is typed as ParsedResume|null
// rather than a loose object - it's exactly that shape once status is
// 'completed' (see ParsedResumeDisplay.vue, the one place this gets
// rendered), null otherwise.
export interface ResumeAnalysis {
  id: string
  document_id: string
  status: AIJobStatus
  parsed_data: ParsedResume | null
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface ResumeAnalysisListResponse {
  items: ResumeAnalysis[]
  total: number
  page: number
  page_size: number
}

export interface ResumeAnalysisListParams {
  document_id?: string
  page?: number
  page_size?: number
}

// --- ATS scores ----------------------------------------------------------

export interface AtsScoreCreatePayload {
  resume_analysis_id: string
  application_id?: string | null
  /** Optional - prefers the linked application's job_url when omitted.
   * Server-enforced 50-20000 char bounds when provided; see
   * AtsScoresView.vue's create dialog for the client-side hint. */
  job_description?: string | null
}

// Mirrors AtsScoreRead. `feedback` is typed as AtsScoreResult|null for the
// same reason ResumeAnalysis.parsed_data is - exact shape once completed.
export interface AtsScore {
  id: string
  resume_analysis_id: string
  application_id: string | null
  job_description: string | null
  job_description_source: 'pasted' | 'url' | null
  status: AIJobStatus
  score: number | null
  feedback: AtsScoreResult | null
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface AtsScoreListResponse {
  items: AtsScore[]
  total: number
  page: number
  page_size: number
}

export interface AtsScoreListParams {
  application_id?: string
  page?: number
  page_size?: number
}

File: backend/app/schemas/ai.py
"""
Schemas for the AI features (Resume Parser + ATS Score - TODO.md "AI
Features"). Two different jobs here:

- ParsedResume (+ WorkExperienceItem, EducationItem) and AtsScoreResult
  are both the `response_schema` handed to Gemini (see
  app/services/ai/client.py::call_structured) *and* the validator for its
  response - Gemini's structured-output mode fills in exactly this shape,
  and re-validating it here (rather than trusting the SDK blindly) is
  what lets a malformed/incomplete response surface as a real failure.
  Every field on these four models is required (`Field(...)`, no Python-
  side default) even where the type itself is nullable (`str | None`) -
  Gemini's API rejects a response_schema containing default values
  outright ("Default value is not supported in the response schema for
  the Gemini API" - confirmed against googleapis/python-genai#699), so
  "optional" here is expressed purely as "the key must be present, but
  its value may be null/empty", never as "the key may be omitted". Field
  types are otherwise kept deliberately simple (str, str | None,
  list[str], list[<flat submodel>]) - Gemini's response_schema support is
  an OpenAPI subset, not arbitrary Pydantic (no $ref/oneOf/allOf/
  recursive schemas).
- ResumeAnalysisCreate/Read/ListResponse and AtsScoreCreate/Read/
  ListResponse are the ordinary request/response schemas for
  app/api/v1/endpoints/ai.py, same pagination-envelope shape as every
  other list endpoint (e.g. DocumentListResponse). These aren't sent to
  Gemini, so the "no defaults" constraint above doesn't apply to them.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.resume_analysis import AIJobStatus


# --- Gemini structured-output schemas --------------------------------------


class WorkExperienceItem(BaseModel):
    company: str = Field(...)
    title: str = Field(...)
    start_date: str | None = Field(...)
    end_date: str | None = Field(...)  # "Present" (as text) if currently held
    description: str | None = Field(...)


class EducationItem(BaseModel):
    institution: str = Field(...)
    degree: str | None = Field(...)
    field_of_study: str | None = Field(...)
    graduation_date: str | None = Field(...)


class ParsedResume(BaseModel):
    full_name: str | None = Field(...)
    email: str | None = Field(...)
    phone: str | None = Field(...)
    summary: str | None = Field(...)
    skills: list[str] = Field(...)
    work_experience: list[WorkExperienceItem] = Field(...)
    education: list[EducationItem] = Field(...)
    total_years_experience: float | None = Field(...)


class AtsScoreResult(BaseModel):
    score: int = Field(..., ge=0, le=100)
    summary: str = Field(...)
    matched_keywords: list[str] = Field(...)
    missing_keywords: list[str] = Field(...)
    suggestions: list[str] = Field(...)


# --- Resume analyses API ----------------------------------------------------


class ResumeAnalysisCreate(BaseModel):
    document_id: uuid.UUID


class ResumeAnalysisRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    status: AIJobStatus
    parsed_data: dict | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class ResumeAnalysisListResponse(BaseModel):
    items: list[ResumeAnalysisRead]
    total: int
    page: int
    page_size: int


# --- ATS scores API ----------------------------------------------------------


class AtsScoreCreate(BaseModel):
    resume_analysis_id: uuid.UUID
    application_id: uuid.UUID | None = None
    # Optional: prefers the linked application's job_url when omitted -
    # see app/api/v1/endpoints/ai.py's create endpoint and
    # app/tasks/ai.py::score_ats_task. Length-bounded so an empty/garbage
    # submission doesn't burn a real Gemini call.
    job_description: str | None = Field(default=None, min_length=50, max_length=20000)


class AtsScoreRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    resume_analysis_id: uuid.UUID
    application_id: uuid.UUID | None
    job_description: str | None
    job_description_source: str | None
    status: AIJobStatus
    score: int | None = None
    feedback: dict | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class AtsScoreListResponse(BaseModel):
    items: list[AtsScoreRead]
    total: int
    page: int
    page_size: int
